### ISSUE TITLE
Add scalar fallback path tests for gradient clipping

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5145]
The SIMD gradient clipping functions have scalar fallback paths for non-float dtypes (e.g., int32). These fallback paths in `compute_gradient_norm_list`, `clip_gradients_by_global_norm`, `clip_gradients_per_param`, and `clip_gradients_by_value_list` are not covered by any tests. While less critical than the SIMD paths, they should be tested for correctness.

_Follow-up from #4911_

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have all the context I need. Let me create the implementation plan.

---

# Implementation Plan: Issue #5145 — Scalar Fallback Path Tests for Gradient Clipping

## 1. Objective

Add test coverage for the scalar fallback paths (non-float32/float64 dtypes) in four gradient clipping functions:

- `compute_gradient_norm_list` — scalar norm accumulation (lines 187–192)
- `clip_gradients_by_global_norm` — scalar scaling after norm (lines 246–250)
- `clip_gradients_per_param` — scalar norm + scalar scaling (lines 293–314)
- `clip_gradients_by_value_list` — scalar clamp (lines 359–367)

The `else:` branches in each function are currently exercised only by float32/float64 tests, which take the SIMD paths. Using `DType.int32` will route execution through these scalar fallback branches.

---

## 2. Approach

- **Single file modification**: Add 4 new test functions directly to the existing test file `tests/shared/training/test_gradient_clipping.mojo`. No new file needed.
- **Dtype**: Use `DType.int32` — it is not float32 or float64, so it unconditionally hits all `else:` branches.
- **Assertions**: Exact integer equality checks — no floating-point tolerance needed since the values flow through `_get_float64`/`_set_float64` but the original values are small integers with deterministic results.
- **Test values**: Small, non-uniform integers (e.g., `[1, 2, 3, 4]`) that make norm and clipping outcomes easy to verify by hand.
- **Main function**: Update the numbered test runner to include 4 new tests (18–21), update the total count from 17 to 21.

---

## 3. Files to Create

None. All changes go into the existing test file.

---

## 4. Files to Modify

### `tests/shared/training/test_gradient_clipping.mojo`

**Add 4 test functions** after the existing float64 tests (after line 585, before `def main()`):

#### `test_compute_gradient_norm_scalar_fallback()` (new, ~25 lines)
- Create one `AnyTensor` with `DType.int32`, values `[1, 2, 3, 4]`
- Call `compute_gradient_norm_list([grad])`
- Expected norm: `sqrt(1² + 2² + 3² + 4²) = sqrt(30) ≈ 5.477`
- Assert with `assert_close_float` using `atol=1e-4` (Float32 return, slight sqrt rounding)
- Verify values are unchanged afterward

#### `test_clip_by_global_norm_scalar_fallback()` (new, ~40 lines)
- Create `DType.int32` tensor with values `[3, 4]` → norm = `sqrt(9+16) = 5.0`
- Call `clip_gradients_by_global_norm(grads, max_norm=Float32(2.5))`
- Returns pre-clip norm `5.0`; `clip_coef = 2.5 / (5.0 + 1e-6) ≈ 0.5`
- Expected scaled values: `[3 * 0.5, 4 * 0.5]` → stored as float64 in int32 tensor → `[1, 2]` (truncated)
- Assert return value close to `5.0`
- Assert post-clip values via `_get_float64`
- Also test the no-clipping case: norm ≤ max_norm → values unchanged

#### `test_clip_per_param_scalar_fallback()` (new, ~40 lines)
- Create two `DType.int32` tensors: `[3, 4]` (norm=5) and `[1, 1]` (norm≈1.414)
- Call `clip_gradients_per_param(grads, max_norm=Float32(2.5))`
- First tensor: norm 5 > 2.5 → clipped; second tensor: norm 1.414 ≤ 2.5 → unchanged
- Verify first tensor values scale by `2.5/5 = 0.5`: `[1.5, 2.0]`
- Verify second tensor values unchanged: `[1.0, 1.0]`

#### `test_clip_by_value_scalar_fallback()` (new, ~35 lines)
- Create `DType.int32` tensor with values `[-3, -1, 0, 1, 3]`
- Call `clip_gradients_by_value_list(grads, min_value=-2.0, max_value=2.0)`
- Expected: `[-2, -1, 0, 1, 2]` (clamped at boundaries)
- Assert each element via `_get_float64`
- Also verify in-range values are completely unchanged

**Update `def main()`:**
- Change total from 17 to 21
- Add 4 new print/call/assert blocks for tests 18–21
- Update final summary line: `"All 21 gradient clipping tests PASSED!"`

---

## 5. Implementation Order

1. **Read** `tests/shared/training/test_gradient_clipping.mojo` lines 440–586 to understand the float64 test pattern to mirror
2. **Read** the imports at the top of the test file to confirm `AnyTensor`, `full`, `ones`, `zeros` are already imported
3. **Write** `test_compute_gradient_norm_scalar_fallback()` immediately after line 585 (the last float64 test)
4. **Write** `test_clip_by_global_norm_scalar_fallback()`
5. **Write** `test_clip_per_param_scalar_fallback()`
6. **Write** `test_clip_by_value_scalar_fallback()`
7. **Edit** `def main()` to register all 4 new tests (tests 18–21) and update total counts
8. **Run** `mojo run tests/shared/training/test_gradient_clipping.mojo` to verify all 21 pass
9. **Commit** and create PR linked to #5145

---

## 6. Verification

```bash
# Run locally (if GLIBC ≥ 2.32)
mojo run tests/shared/training/test_gradient_clipping.mojo

# Or via container
just shell
mojo run tests/shared/training/test_gradient_clipping.mojo

# Or via justfile test group
just test-group "tests/shared/training" "test_gradient_clipping.mojo"
```

**Expected output:**
```
[18/21] Testing scalar fallback gradient norm computation...
✓ PASSED
[19/21] Testing scalar fallback global norm clipping...
✓ PASSED
[20/21] Testing scalar fallback per-param clipping...
✓ PASSED
[21/21] Testing scalar fallback value clipping...
✓ PASSED

All 21 gradient clipping tests PASSED! ✓
```

**Manual verification of test values:**

| Test | Input | Operation | Expected |
|------|-------|-----------|----------|
| Norm | `[1,2,3,4]` int32 | `sqrt(1+4+9+16)` | `≈5.477` |
| Global clip | `[3,4]` int32, max=2.5 | scale by `2.5/5.0=0.5` | `[1.5, 2.0]` |
| Per-param clip | `[3,4]`, `[1,1]` int32, max=2.5 | first clipped, second unchanged | `[1.5,2.0]`, `[1.0,1.0]` |
| Value clip | `[-3,-1,0,1,3]` int32, range=[-2,2] | clamp | `[-2,-1,0,1,2]` |

---

## 7. Key Implementation Notes

- **`_get_float64` on int32**: Returns the integer value as Float64 (e.g., `3` → `3.0`). The scalar paths rely on this for computation.
- **`_set_float64` on int32**: Stores the float64 value back into the int32 storage. For values like `1.5`, this will truncate to `1`. Use `_get_float64` (not integer equality) to check post-clip values to avoid truncation surprises — or choose test values where the result is exact (e.g., `[2, 4]` with `clip_coef=0.5` → `[1.0, 2.0]`).
- **Prefer exact results**: Pick input values and `max_norm` such that `input * clip_coef` is a whole number. E.g., `[2, 4]` with norm `sqrt(20)≈4.47`, max_norm=`2.0` gives `clip_coef≈0.447` → not clean. Better: use `[3, 4]` (norm=5) with max_norm=2.5 → `clip_coef=0.5` exactly → `[1.5, 2.0]` stored as float64 in int32 slot → read back as `1.5, 2.0` via `_get_float64`. Assert with small atol.
- **No new imports needed**: The existing test file already imports everything required (`AnyTensor`, `full`, `zeros`, `assert_close_float`, `assert_equal_int`, `assert_true`).

---

## 8. Skills Used

### Invoked During Planning
- `hephaestus:advise` — searched team knowledge base for prior learnings before planning
- `Explore` subagent — codebase exploration to read implementation and test structure

### Team Knowledge Base Skills Referenced (from Prior Learnings)
- `fix-ci-mojo-test-failure` — Mojo test execution patterns; confirmed `mojo run` (not `mojo test`) is correct
- `batch-norm-backward-gradient-analysis` — gradient field name conventions (not directly applicable here, but used for orientation)
- `batch-norm-gradient-test-fix` — non-uniform gradient patterns to avoid false positives; confirmed using `[1,2,3,4]` over `ones()`
- `mojo-dtype-partial-support-guard` — verified int32 `_get_float64`/`_set_float64` behaves correctly (unlike BF16 which silently zeros)
- `dtype-native-migration` — DType mapping strategies; confirmed int32 is the right choice for scalar fallback coverage

---
*Generated by Claude Code Planner*
